addons/sale/models/account_move.py

```diff
@@ -0,0 +1,190 @@
+                move_line_id = values.get('move_id')
+                sale_line = map_sale_line_per_move.get(move_line_id)
+                if sale_line:
+                    values['so_line'] = sale_line.id
+
+        return values_list
+
+    @api.multi
+    def _sale_create_reinvoice_sale_line(self):
+
+        sale_order_map = self._sale_determine_order()
+
+        sale_line_values_to_create = []  # the list of creation values of sale line to create.
+        existing_sale_line_cache = {}  # in the sales_price-delivery case, we can reuse the same sale line. This cache will avoid doing a search each time the case happen
+        # `map_move_sale_line` is map where
+        #   - key is the move line identifier
+        #   - value is either a sale.order.line record (existing case), or an integer representing the index of the sale line to create in
+        #     the `sale_line_values_to_create` (not existing case, which will happen more often than the first one).
+        map_move_sale_line = {}
+
+        for move_line in self:
+            sale_order = sale_order_map.get(move_line.id)
+
+            # no reinvoice as no sales order was found
+            if not sale_order:
+                continue
+
+            # raise if the sale order is not currenlty open
+            if sale_order.state != 'sale':
+                message_unconfirmed = _('The Sales Order %s linked to the Analytic Account %s must be validated before registering expenses.')
+                messages = {
+                    'draft': message_unconfirmed,
+                    'sent': message_unconfirmed,
+                    'done': _('The Sales Order %s linked to the Analytic Account %s is currently locked. You cannot register an expense on a locked Sales Order. Please create a new SO linked to this Analytic Account.'),
+                    'cancel': _('The Sales Order %s linked to the Analytic Account %s is cancelled. You cannot register an expense on a cancelled Sales Order.'),
+                }
+                raise UserError(messages[sale_order.state] % (sale_order.name, sale_order.analytic_account_id.name))
+
+            price = move_line._sale_get_invoice_price(sale_order)
+
+            # find the existing sale.line or keep its creation values to process this in batch
+            sale_line = None
+            if move_line.product_id.expense_policy == 'sales_price' and move_line.product_id.invoice_policy == 'delivery':  # for those case only, we can try to reuse one
+                map_entry_key = (sale_order.id, move_line.product_id.id, price)  # cache entry to limit the call to search
+                sale_line = existing_sale_line_cache.get(map_entry_key)
+                if sale_line:  # already search, so reuse it. sale_line can be sale.order.line record or index of a "to create values" in `sale_line_values_to_create`
+                    map_move_sale_line[move_line.id] = sale_line
+                    existing_sale_line_cache[map_entry_key] = sale_line
```

Comment:
As you can easily understand with the comment, you can directly assign both values
`map_move_sale_line[move_line.id] = existing_sale_line_cache[map_entry_key] = sale_line`